### PR TITLE
8213269: convert test/hotspot/jtreg/runtime/memory/RunUnitTestsConcurrently to gtest

### PR DESCRIPTION
**Pull request by mseledts**
This is a preliminary review. I would like to get the initial feedback before I proceed with conversion of the remaining tests.

Here is what I did so far:
  - created a UnitTestThread and a main test runner, based on gtests with similar needs
  - moved the original code from HotSpot internals (so called hotspot internal tests: src/hotspot/share/memory/virtualspace.cpp)
    to the newly created gtest while wrapping it into a TestReservedSpace class. I did not change the code of the test.
  - removed invocations from whitebox.cpp

Testing:
  - ran GTestWrapper on usual platforms - All PASS
  - ensured that ReservedSpaceConcurrent is in the logs and passed

After gathering the feedback my plan is:
Plan:
  - move the remaining internal Memory/VirtualSpace tests into a gTest
    - I am thinking about using separate files for each test
  - create a common file for UnitTestThread and MultiThreadTestRunner to reuse the code

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8213269](https://bugs.openjdk.java.net/browse/JDK-8213269): convert test/hotspot/jtreg/runtime/memory/RunUnitTestsConcurrently to gtest

### Reviewers
 * [Igor Ignatyev](https://openjdk.java.net/census#iignatyev) (@iignatev - **Reviewer**) ⚠️ Review applies to d9f618c2ec893f186670c4ef5f416c04218662ae
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2436/head:pull/2436`
`$ git checkout pull/2436`

### CODE REVIEW

**Comment by mseledts** [comment]
@iignatev @hseigel Could you, please, take a look when you have a chance? This is a preliminary review.

File: test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    long stopTime = os::javaTimeMillis() + TEST_DURATION;
+    while (os::javaTimeMillis() < stopTime) {
+      run_unit_test();
+    }
+  }
+
+private:
+  void run_unit_test() {
+    TestReservedSpace::test_reserved_space();
+  }
```
Comment (iignatev) [inline test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp:208]: what's the point of having this extra member function? can't we just call `TestReservedSpace::test_reserved_space` directly from main_run?

File: test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+    char* end = (char*)addr + size;
+    for (char* p = (char*)addr; p < end; p += page_size) {
+      *p = 1;
+    }
+  }
+
+  static void release_memory_for_test(ReservedSpace rs) {
+    if (rs.special()) {
+      guarantee(os::release_memory_special(rs.base(), rs.size()), "Shouldn't fail");
```
Comment (iignatev) [inline test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp:46]: you shouldn't use hotspot's `gurantee/assert`, and use gtest provided `ASSERT`/`EXPECT` instead.

**Review by iignatev — CHANGES_REQUESTED** [review]
hi Misha,

I haven't finished (pre)reviewing it yet, however, I have found a few things that, I belive, should be changed.

File: test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp (hunk 2 above)
Comment (mseledts) [reply test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp]: Will do.

File: test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp (hunk 1 above)
Comment (mseledts) [reply test/hotspot/gtest/runtime/test_reservedSpaceConcurrent.cpp]: I reworked the classes to be able to reuse the concurrency driver and threads code.

**Comment by mseledts** [comment]
I have updated the changes with recommendations from Igor. I also have:
  - converted all 3 tests, which also includes platforms-specific tests
  - using GTest EXPECT/ASSERT macros
  - removed remnants of the original tests
  - general cleanup
  - ran build-and-test on these tests, plus a number of additional builds
Please review when you have a chance: @iignatev @hseigel

**Review by coleenp — APPROVED** [review]
Looks good! Thank you for making this a gtest.

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+#define GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
+
+#include "threadHelper.inline.hpp"
+
+// This file contains helper classes to run unit tests concurrently in multiple threads.
+
+// Base class for test runnable. Override runUnitTest() to specify what to run.
+class TestRunnable {
+public:
+  virtual void runUnitTest() {
```
Comment (iignatev) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:34]: I assume you meant for this class to be abstract, if so `runUnitTest` should be a pure virtual function.

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/*
```
Comment (iignatev) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:1]: for c++, we don't use camelCase in filenames, but rather use small_snake_case

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    for (int i = 0; i < nrOfThreads; i++) {
+      done.wait();
+    }
+  }
+
+private:
+  long testDurationMillis;
+  int nrOfThreads;
+  TestRunnable* unitTestRunnable;
```
Comment (iignatev) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:97]: these also can be made `const`

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  // from JavaTestThread
+  void main_run() {
+    long stopTime = os::javaTimeMillis() + testDuration;
+    while (os::javaTimeMillis() < stopTime) {
+      runnable->runUnitTest();
+    }
+  }
+private:
+  long testDuration;
+  TestRunnable* runnable;
```
Comment (iignatev) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:60]: all these can be made `const` and initialized in the initializer list.

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+public:
+  // runnableArg - what to run
+  // doneArg - a semaphore to notify when the thread is done running
+  // testDurationArg - how long to run (in milliseconds)
+  UnitTestThread(TestRunnable* runnableArg, Semaphore* doneArg, long testDurationArg) : JavaTestThread(doneArg) {
+    runnable = runnableArg;
+    testDuration = testDurationArg;
+  }
+
+  virtual ~UnitTestThread() {}
```
Comment (iignatev) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:49]: why do you need virtual d-ctor here?

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  }
+
+private:
+  long testDurationMillis;
+  int nrOfThreads;
+  TestRunnable* unitTestRunnable;
+};
+
+#endif // include guard
```
Comment (iignatev) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:100]: we tend to use the expression used in the corresponding `#if` / `#ifdef` as a comment in `#endif`

**Review by iignatev — CHANGES_REQUESTED** [review]

File: test/hotspot/gtest/runtime/test_os_windows.cpp [hunk 9]
```diff
@@ -688,4 +689,20 @@ TEST_VM(os_windows, handle_long_paths) {
...
+
+class ReserveMemorySpecialRunnable : public TestRunnable {
+public:
+  void runUnitTest() {
+    TestReserveMemorySpecial_test();
+  }
+};
+
+TEST_VM(os_windows, reserve_memory_special_concurrent) {
+  ConcurrentTestRunner testRunner(new ReserveMemorySpecialRunnable(), 30, 15000);
```
Comment (iignatev) [inline test/hotspot/gtest/runtime/test_os_windows.cpp:704]: a memory leak

File: test/hotspot/gtest/runtime/test_os_linux.cpp [hunk 10]
```diff
@@ -242,4 +243,179 @@ TEST_VM(os_linux, reserve_memory_special_shm) {
...
+
+class ReserveMemorySpecialRunnable : public TestRunnable {
+public:
+  void runUnitTest() {
+    TestReserveMemorySpecial::test();
+  }
+};
+
+TEST_VM(os_linux, reserve_memory_special_concurrent) {
+  ConcurrentTestRunner testRunner(new ReserveMemorySpecialRunnable(), 30, 15000);
```
Comment (iignatev) [inline test/hotspot/gtest/runtime/test_os_linux.cpp:417]: yet another memory leak

File: test/hotspot/gtest/memory/test_virtualspace.cpp [hunk 11]
```diff
@@ -337,3 +338,341 @@ TEST_VM(VirtualSpace, disable_large_pages) {
...
+
+class VirtualSpaceRunnable : public TestRunnable {
+public:
+  void runUnitTest() {
+    TestVirtualSpace::test_virtual_space();
+  }
+};
+
+TEST_VM(VirtualSpace, virtual_space_concurrent) {
+  ConcurrentTestRunner testRunner(new VirtualSpaceRunnable(), 30, 15000);
```
Comment (iignatev) [inline test/hotspot/gtest/memory/test_virtualspace.cpp:676]: one more memory leak

File: test/hotspot/gtest/memory/test_virtualspace.cpp [hunk 12]
```diff
@@ -337,3 +338,341 @@ TEST_VM(VirtualSpace, disable_large_pages) {
...
+
+class ReservedSpaceRunnable : public TestRunnable {
+public:
+  void runUnitTest() {
+    TestReservedSpace::test_reserved_space();
+  }
+};
+
+TEST_VM(VirtualSpace, reserve_space_concurrent) {
+  ConcurrentTestRunner testRunner(new ReservedSpaceRunnable(), 30, 15000);
```
Comment (iignatev) [inline test/hotspot/gtest/memory/test_virtualspace.cpp:664]: and a memory leak again

**Review by iignatev — CHANGES_REQUESTED** [review]

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp (hunk 4 above)
Comment (mseledts) [reply test/hotspot/gtest/concurrentTestRunner.inline.hpp:1]: OK. I saw gtestMain.cpp, gtestLauncher.cpp and a few others, and just followed that. I also see a number of test_camelCase.cpp: test_primitiveConversions.cpp, test_logSelectionList.cpp and so on. In fact, it seems the most prevalent pattern for gtests is test_camelCase.cpp.

Anyway, no problem, I can rename this file to concurrent_test_runner.inline.hpp

File: test/hotspot/gtest/runtime/test_os_windows.cpp (hunk 9 above)
Comment (mseledts) [reply test/hotspot/gtest/runtime/test_os_windows.cpp:704]: I forgot I am not in Java anymore :)

**Comment by mseledts** [comment]
I believe I have addressed the 2nd round of review feedback. @iignatev Igor, could you please take a look when you have a chance?

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp [hunk 13]
```diff
@@ -92,9 +86,9 @@ class ConcurrentTestRunner {
   }
 
 private:
-  long testDurationMillis;
-  int nrOfThreads;
-  TestRunnable* unitTestRunnable;
+  const TestRunnable* unitTestRunnable;
```
Comment (iignatev) [inline test/hotspot/gtest/concurrent_test_runner.inline.hpp:89]: you made it a pointer to const TestRunnable, not a const pointer to TestRunnable.
```suggestion
  TestRunnable* const unitTestRunnable;
```

File: test/hotspot/gtest/memory/test_virtualspace.cpp [hunk 14]
```diff
@@ -655,24 +655,28 @@ class TestVirtualSpace : AllStatic {
...
     TestVirtualSpace::test_virtual_space();
   }
 };
 
 TEST_VM(VirtualSpace, virtual_space_concurrent) {
-  ConcurrentTestRunner testRunner(new VirtualSpaceRunnable(), 30, 15000);
+  TestRunnable* runnable = new VirtualSpaceRunnable();
+  ConcurrentTestRunner testRunner(runnable, 30, 15000);
   testRunner.run();
+  delete runnable;
```
Comment (iignatev) [inline test/hotspot/gtest/memory/test_virtualspace.cpp]: wouldn't it be easier to allocate TestRunnable on stack and pass a pointer?
```suggestion
  VirtualSpaceRunnable runnable();
  ConcurrentTestRunner testRunner(&runnable, 30, 15000);
  testRunner.run();
```

**Review by iignatev — CHANGES_REQUESTED** [review]

File: test/hotspot/gtest/memory/test_virtualspace.cpp (hunk 14 above)
Comment (mseledts) [reply test/hotspot/gtest/memory/test_virtualspace.cpp]: Thanks Igor for these suggestions. I have updated the code accordingly.

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+// Helper class for running a given unit test concurrently in multiple threads.
+class ConcurrentTestRunner {
+public:
+  // runnableArg - what to run
+  // nrOfThreadsArg - how many threads to use concurrently
+  // testDurationMillisArg - duration for each test run
+  ConcurrentTestRunner(TestRunnable* const runnableArg, int nrOfThreadsArg, long testDurationMillisArg) :
+    unitTestRunnable{runnableArg},
+    nrOfThreads{nrOfThreadsArg},
+    testDurationMillis{testDurationMillisArg} {}
```
Comment (iignatev) [inline test/hotspot/gtest/concurrent_test_runner.inline.hpp:67]: why did you use `{}` instead of `()` to init these (as well as UnitTestThread's) fields?

**Review by iignatev — APPROVED** [review]

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi,

Have we decided that the STL can be used now? If yes, I must have missed this. If no, could you please change the ConcurrentTestRunner to use GrowableArray or just a simply a malloced array?

Thanks!

..Thomas

**Comment by mseledts** [comment]
@tstuefe @iignatev 
On a question: Have we decided that the STL can be used now? 

Hi Thomas,
  I do most of my JDK work in Java, hence may have missed knowing the restriction of using STL.
I wonder if this restriction is only for source code, and not the tests. I did a quick search under "test/hotspot/gtest", and found many uses of std::, including the data structures. For instance, jfr/test_networkUtilization.cpp uses std::map, std::list and std::vector.

Thanks,
Misha

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    unitTestRunnable(runnableArg),
+    nrOfThreads(nrOfThreadsArg),
+    testDurationMillis(testDurationMillisArg) {}
+
+  virtual ~ConcurrentTestRunner() {}
+
+  void run() {
+    Semaphore done(0);
+
+    std::vector<UnitTestThread*> t;
```
Comment (coleenp) [inline test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: Regarding the STL question, I think using std in a limited fashion in the tests seems ok before we start allowing it in the main sources, despite where Misha pointed out it already exists.  It's going to be allowed pretty soon anyway.
One question about this: where does it allocate memory for 't' ?  Are the elements leaked here?

Comment (mseledts) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: On a question: where does it allocate memory for 't' ? Are the elements leaked here?
The vector will allocate and free the memory for the containers (the memory containing pointers to UnitTestThread). 

As for the UnitTestThread objects/instances themselves:
- JavaTestThread. I checked couple of other examples of uses of JavaTestThread (test_symbolTable.cpp and oops/test_markWord.cpp), and neither is calling explicit delete on the objects of this type
- I presume the thread will free its resources once it exits (and we wait for done semafore to ensure this)
- adding code “delete t[i]” results in crash, presuming due to freeing an already released memory

Hence, I believe, no changes needed to this code to free resources.

Comment (tstuefe) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: > Regarding the STL question, I think using std in a limited fashion in the tests seems ok before we start allowing it in the main sources, despite where Misha pointed out it already exists. It's going to be allowed pretty soon anyway.

Really? I recently had a private discussion with Kim Barret about this and understood that this is a contentious point. I was hoping we would have a public discussion about this before deciding on this. My fears are increased build times (which seem to get worse and worse) and stability- and compiler issues. I am not completely against it, but I'm a bit of a burned child wrt to STL.

Comment (tstuefe) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: > One question about this: where does it allocate memory for 't' ? Are the elements leaked here?

Backing buffer for vector lives in C-heap. Its a dynamically growing array, basically like our GrowableArray. Only its not under our control and we cannot easily debug it if something goes wrong (take a look at the STL sources, they are not easy to understand). And it sidesteps our own os::malloc, so we cannot account it, and our guards won't work.

Personally I would like to keep the test harness as simple stupid as possible, to avoid interfering with the actual test. Especially with tests which test memory allocation. I know that is a vague reason, but I also cannot find a bit advantage in using std::vector, compared with a simple calloced array. We dont need the dynamically-growing part here, we know right upfront how many threads we start, so its really a fixed sized array.

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
+#define GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
+
+#include "threadHelper.inline.hpp"
```
Comment (tstuefe) [inline test/hotspot/gtest/concurrent_test_runner.inline.hpp:27]: Make sure you include all headers needed for this file. Includes should be self-contained, so pull everything they need (basically, you should be able to include it into an empty cpp file and it should build fine). You use Semaphore and some os::xxx functions, so you'd need at least os.hpp and wherever Semaphore lives.

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+public:
+  // runnableArg - what to run
+  // nrOfThreadsArg - how many threads to use concurrently
+  // testDurationMillisArg - duration for each test run
+  ConcurrentTestRunner(TestRunnable* const runnableArg, int nrOfThreadsArg, long testDurationMillisArg) :
+    unitTestRunnable(runnableArg),
+    nrOfThreads(nrOfThreadsArg),
+    testDurationMillis(testDurationMillisArg) {}
+
+  virtual ~ConcurrentTestRunner() {}
```
Comment (tstuefe) [inline test/hotspot/gtest/concurrent_test_runner.inline.hpp:69]: Do we derive from this class? And delete via base pointers? If not, I'd remove this.

**Review by tstuefe — COMMENTED** [review]
Hi Misha,

> Hi Thomas,
> I do most of my JDK work in Java, hence may have missed knowing the restriction of using STL.
> I wonder if this restriction is only for source code, and not the tests. I did a quick search under "test/hotspot/gtest", and found many uses of std::, including the data structures. For instance, jfr/test_networkUtilization.cpp uses std::map, std::list and std::vector.

See my answer to Coleen. I think using the STL would have a number of repercussions which should be discussed before doing this step. 

About the patch itself:

I am not sure moving more and more test control down into gtest is a good thing. gtests have mostly be single threaded until now and could be run within a make. AFAIU gtests offer way less run control than jtreg does, e.g. you cannot disable the test without recompiling the hotspot (there is no ProblemList equivalent for gtest), you have no test groupings (e.g. to separate stress tests which need a whole machine for themselves) etc. Also, we will duplicate more and more thread control in C++.

I understand the wish to remove the test coding from the hotspot implementation. Its so ugly and should not live there. But just moving them to separate implementation files, possibly within a clearly marked "tests" folder, would be a first good step.

I will not block this if you have decided to go this way. Just wanted to understand the direction you guys plan to go with gtests in the future. If they get more complex and powerful we may need more control, eg the possibility to problemlist tests.

Cheers, Thomas

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp (hunk 16 above)
Comment (coleenp) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: Okay so you're right that we can't account in NMT and with our malloc guards for the memory that is allocated with std::vector, so I'm convinced that Misha should change this.  Hopefully we can figure out how to make std::vector work in the future, because Misha's 2 lines for it look quite nice.
Do you think using std:: would increase build times? My suspicion is that the Access barrier templates are to blame, but building hotspot is a fraction of building the whole jdk, which does seem to take a long time.

**Comment by coleenp** [comment]
Moving these tests out of our source code is a good change.  We already have a couple multithreaded tests in gtest, and adding this one isn't going to make the gtests unreliable.  For future multithreaded tests, it makes sense to make them jtreg tests so that we can control their execution.  So I'd like to see this PR integrated, with suggested changes.

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp (hunk 4 above)
Comment (kimbarrett) [reply test/hotspot/gtest/concurrentTestRunner.inline.hpp]: This statement that we use small_snake_case for filenames is mostly wrong.  Most HotSpot files contain some primary class, whose name is CamelCased, and the associated files are camelCased.  There are a fairly small number of exceptions to that.  There are some suffixes and prefixes that get added on, such as cpu and os suffixes (`_linux`, `_windows`, &etc, suffixes, `test_` gtest prefix, component prefixes like `c1_` and `gc_`), but those are mostly add-ons to the basic camelCase convention.

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp (hunk 16 above)
Comment (tstuefe) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: > Okay so you're right that we can't account in NMT and with our malloc guards for the memory that is allocated with std::vector, so I'm convinced that Misha should change this. Hopefully we can figure out how to make std::vector work in the future, because Misha's 2 lines for it look quite nice.
> Do you think using std:: would increase build times? My suspicion is that the Access barrier templates are to blame, but building hotspot is a fraction of building the whole jdk, which does seem to take a long time.

Well, yesterday I saw to my annoyance that upgrading from g++-7 to g++-9 increased jdk build time by about 30-40%, so I today I am very aware of build times :-/ But you are right of course, its just the hotspot, so maybe its not such a big deal.

I am a bit apprehensive about introducing STL (had to port the STL to outlier platforms in the past and god that was terrible) but mainly I'd like to have a public discussion before starting to use it. And if the majority thinks its okay then I am okay with it too oc. I have been wrong before. Kim said STL quality is very high nowadays.

**Comment by tstuefe** [comment]
> Moving these tests out of our source code is a good change. 

Yes I agree. The tests were an eyesore.

> We already have a couple multithreaded tests in gtest, and adding this one isn't going to make the gtests unreliable. For future multithreaded tests, it makes sense to make them jtreg tests so that we can control their execution. So I'd like to see this PR integrated, with suggested changes.

Okay then. I am fine with this change too. I recently learned that tests can be disabled in the source code by preceding them with "DISABLED_". Not as good as ProblemLists, since it requires recompilation, but better than nothing.

---
@mseledts :

Would it be possible to rename the gtests to make them start with "os"? We recently added jtreg tests which run the "os" part of the gtests with various large page options (see https://bugs.openjdk.java.net/browse/JDK-8257959), and if you name your tests "os" too they get tested automatically with various large page settings. Which would be a nice benefit from making them gtests.

Cheers, Thomas

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp (hunk 4 above)
Comment (iignatev) [reply test/hotspot/gtest/concurrentTestRunner.inline.hpp]: apparently, I had a false memory, we indeed have more `camelCase` than `small_snake_case` files, so @mseledts, you will need to revert that. sorry for wasting your time. in my defense, most of the files are just `case` hence can be interpreted either way :)

Comment (mseledts) [reply test/hotspot/gtest/concurrentTestRunner.inline.hpp]: No worries. I will 'git rename' it back to the original name, I hope this will work.

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp (hunk 16 above)
Comment (mseledts) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: Thanks for your feedback and discussion. I will refrain from using STL, and use malloc to allocate the array. I presume I should use os::malloc() in this case - please correct me if I am wrong.

**Comment by iignatev** [comment]
Hi Thomas,

> <... >Just wanted to understand the direction you guys plan to go with gtests in the future. If they get more complex and powerful we may need more control, eg the possibility to problemlist tests. <...>

as you have discovered yourself, gtest already has support for the temporary exclusion of tests. I agree that it inferior to problemlists and I'd like to let you know that there is a plan to make it possible to use a problemlist-like solution to exclude gtests. however, it's not on the top of my to-do list as  **currently** there is no big need for that. should it change I'll reprioritize that work. 

-- Igor

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp (hunk 17 above)
Comment (mseledts) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:27]: Thank you for pointing this out. I will make the necessary updates.

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp (hunk 18 above)
Comment (mseledts) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:69]: Not deriving from this class, I will remove the virtual destructor.

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp (hunk 4 above)
Comment (kimbarrett) [reply test/hotspot/gtest/concurrentTestRunner.inline.hpp]: See also "Naming and Grouping" here: https://github.com/openjdk/jdk/blob/master/doc/hotspot-unit-tests.md

File: test/hotspot/gtest/concurrent_test_runner.inline.hpp (hunk 16 above)
Comment (kimbarrett) [reply test/hotspot/gtest/concurrent_test_runner.inline.hpp:74]: re: "there are some configuration and build kludges that were required to make that work." - Thinking about it more, it might be that those kludges were needed even without the jfr uses of the Standard Library.  I think that may have been needed because of uses of the Standard Library by the gtest framework itself.  The jfr tests then took advantage of that support.

**Comment by mseledts** [comment]
On question: "Would it be possible to rename the gtests to make them start with "os"? "
OK. I can rename reserve_space_concurrent to os_reserve_space_concurrent and virtual_space_concurrent to os_virtual_space_concurrent, if I do not hear any objections.

**Comment by mseledts** [comment]
Thank you to all the participants for your feedback, discussion and comments.

To summarize I will make the following changes:
 - rename concurrent_test_runner.inline.hpp to concurrentTestRunner.inline.hpp
 - replace use of std::vector by os:malloced array
 - include all the necessary includes into concurrentTestRunner.inline.hpp so it is self-sufficient
 - remove "virtual ~ConcurrentTestRunner() {}"
 - rename reserve_space_concurrent to os_reserve_space_concurrent and virtual_space_concurrent to os_virtual_space_concurrent

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    unitTestRunnable(runnableArg),
+    nrOfThreads(nrOfThreadsArg),
+    testDurationMillis(testDurationMillisArg) {}
+
+  void run() {
+    Semaphore done(0);
+
+    size_t sz = sizeof(UnitTestThread*) * nrOfThreads;
+    UnitTestThread** t = (UnitTestThread**) os::malloc(sz, mtInternal);
+    memset((void*)t, 0, sz);
```
Comment (tstuefe) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:76]: The memset is not needed.

And if you wanted you could shorten this to one line using:
```
UnitTestThread** t = NEW_C_HEAP_ARRAY(UnitTestThread*, nrOfThreads, mtInternal);
```
which would take care of size calculation and the cast for you. Also, it gives you automatic OOM handling should this ever be a problem.

If you do that, aesthetically it would make sense to change the 'os::free' below to: 'FREE_C_HEAP_ARRAY(UnitTestThread**, t);` though its just a wrapper around os::free.

And you'd need to include memory/allocation.hpp.

If you keep using os::malloc, which would be fine too, you need runtime/os.hpp.

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp (hunk 8 above)
Comment (tstuefe) [reply test/hotspot/gtest/concurrentTestRunner.inline.hpp:99]: "include guard" is self evident. We usually write the name of it, like this:
```
#endif // GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
```

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi Misha,

two small remarks. As I wrote yesterday, I think this is a good cleanup despite my doubts. 

I don't see any GAs running. Can you enable them please? Under "Checks" we should see that all platforms build; also, since the gtests are part of tier1, they run too and we see that they work in all configurations. (eg like this: https://github.com/openjdk/jdk/pull/2751/checks)

You may have to enable github actions if you have never done so and this is your first work in your personal jdk fork (see "Actions" tab under your repo).

Cheers, Thomas

**Comment by mseledts** [comment]
Thomas,

Thank you again for reviewing this change.
  - I will figure out and enable GitHub actions
  - will update allocation/free to use NEW_C_HEAP_ARRAY/FREE_C_HEAP_ARRAY from memory/allocation.hpp
  - #endif // GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
  - will run internal testing, including available builds, tier1 and gtests on all usual platforms
  - if I do not receive additional feedback, my plan is to integrate on Monday

Thanks,
Misha

**Review by tstuefe — APPROVED** [review]
Looks good to me now. I see the GAs did run okay on all platforms, including the gtests (see hotspot tier1 common tests). 

Thanks!

File: test/hotspot/gtest/concurrentTestRunner.inline.hpp [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+  // testDurationMillisArg - duration for each test run
+  ConcurrentTestRunner(TestRunnable* const runnableArg, int nrOfThreadsArg, long testDurationMillisArg) :
+    unitTestRunnable(runnableArg),
+    nrOfThreads(nrOfThreadsArg),
+    testDurationMillis(testDurationMillisArg) {}
+
+  void run() {
+    Semaphore done(0);
+
+    UnitTestThread** t = NEW_C_HEAP_ARRAY(UnitTestThread*, nrOfThreads, mtInternal);
```
Comment (coleenp) [inline test/hotspot/gtest/concurrentTestRunner.inline.hpp:75]: Looks good!

**Review by coleenp — APPROVED** [review]

**Comment by mseledts** [comment]
Thank you again Igor, Coleen, Thomas and Kim for review and discussion.
/integrate
